app/services/fetch_service.rb

```diff
@@ -0,0 +1,451 @@
+    @snapshot.name = 'Fetch [PREPARE_SCORE]'
+    @snapshot.save!
+    score_data = score_snapshot @snapshot, @try
+
+    case_score_manager = CaseScoreManager.new @case
+    case_score_manager.update score_data
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  # rubocop:disable Metrics/MethodLength
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/PerceivedComplexity
+  # rubocop:disable Metrics/BlockLength
+  # rubocop:disable Layout/LineLength
+  def score_snapshot snapshot, try
+    queries_detail = {}
+
+    snapshot.snapshot_queries.each do |snapshot_query|
+      javascript_scorer = JavascriptScorer.new(Rails.root.join('db/scorers/scoring_logic.js'))
+
+      # Prepare some items to score
+      doc_ratings = {}
+      snapshot_query.query.ratings.each do |rating|
+        doc_ratings[rating.doc_id] = rating.rating
+      end
+
+      docs = snapshot_query.snapshot_docs.map do |snapshot_doc|
+        { id: snapshot_doc.doc_id, rating: doc_ratings[snapshot_doc.doc_id] }
+      end
+
+      best_docs = snapshot_query.query.ratings.map do |rating|
+        { id: rating.doc_id, rating: rating.rating }
+      end
+
+      best_docs.sort_by! { |doc| doc[:rating] }.reverse
+
+      # docs = [
+      #  { id: 1, value: 10, rating: 3 },
+      #  { id: 2, value: 20, rating: 0 }
+      # ]
+
+      # Calculate score with options
+      begin
+        scorer = snapshot.scorer
+        code = scorer.code
+
+        score = javascript_scorer.score(docs, best_docs, code)
+        # puts "the score is #{score}"
+        # puts "nan?  #{score.nan?}" if score.is_a? Float
+        if score.is_a?(Float) && score.nan?
+          # snapshot_query.query.notes ||= ""
+          # snapshot_query.query.notes << "\n Fetch Service Snapshot #{snapshot_query.snapshot.id}"
+          # snapshot_query.query.notes << "\n Fetch Service Snapshot Query #{snapshot_query.id}, #{snapshot_query.query.query_text}"
+
+          # snapshot_query.query.notes << "\n Score was NaN"
+          # snapshot_query.query.save!
+          # TODO: Confirm with David Fisher this is right.
+          # Appears to happen when your docs are all rated 0 and your best docs are all 0
+          score = 0
+        end
+        snapshot_query.score = score
+
+        unless snapshot_query.save
+          snapshot_query.query.notes ||= ''
+          snapshot_query.query.notes << "\n Fetch Service Snapshot #{snapshot_query.snapshot.id}"
+          snapshot_query.query.notes << "\n Fetch Service Snapshot Query #{snapshot_query.query.query_text}"
+
+          snapshot_query.query.notes << "\n #{snapshot_query.errors.full_messages}"
+          snapshot_query.query.save!
+        end
+      rescue JavascriptScorer::ScoreError => e
+        puts "Scoring failed: #{e.message}"
+      end
+
+      queries_detail[snapshot_query.query_id] =
+        { score: snapshot_query.score, text: snapshot_query.query.query_text }
+    end
+
+    # Opportunity here to fix the averageing logic
```

Comment:
@david-fisher isn't blindly averaging for the case score NOT the correct thing to do, but it's what we have in Quepid today.   Should we change this up?   Or maybe I am thinking of something else?